source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -418,9 +418,40 @@ void Filter::onComplete(Filters::Common::ExtAuthz::ResponsePtr&& response) {
           *decoder_callbacks_, enumToInt(config_->statusOnError()));
       decoder_callbacks_->streamInfo().setResponseFlag(
           StreamInfo::ResponseFlag::UnauthorizedExternalService);
-      decoder_callbacks_->sendLocalReply(
-          config_->statusOnError(), EMPTY_STRING, nullptr, absl::nullopt,
-          Filters::Common::ExtAuthz::ResponseCodeDetails::get().AuthzError);
+      if (config_->failureDebuggingAllowed()) {
+        ENVOY_STREAM_LOG(trace,
+                         "ext_authz filter has failure debugging allowed. Propagating response "
+                         "body and headers back to the client.",
+                         *decoder_callbacks_);
+        stats_.failure_debugging_allowed_.inc();
+        if (cluster_) {
+          config_->incCounter(cluster_->statsScope(),
+                              config_->ext_authz_failure_debugging_allowed_);
+        }
+        decoder_callbacks_->sendLocalReply(
+            response->status_code, response->body,
+            [&headers = response->headers_to_set,
+             &callbacks = *decoder_callbacks_](Http::HeaderMap& response_headers) -> void {
+              ENVOY_STREAM_LOG(
+                  trace, "ext_authz filter added header(s) to the local response:", callbacks);
+              // Firstly, remove all headers requested by the ext_authz filter, to ensure that they
+              // will override existing headers.
+              for (const auto& header : headers) {
+                response_headers.remove(header.first);
+              }
+              // Then set all the requested headers, allowing the same header to be set multiple
+              // times, e.g. `Set-Cookie`.
+              for (const auto& header : headers) {
+                ENVOY_STREAM_LOG(trace, " '{}':'{}'", callbacks, header.first.get(), header.second);
+                response_headers.addCopy(header.first, header.second);
+              }
```

Comment:
may be `setCopy` is enough?

Reply:
Changed to `setCopy` as suggested. Thanks!